### ISSUE TITLE
improve error handling

### ISSUE THREAD

**Issue by GopherJ** [casbin/casbin-rs#85]
currently we are using:

```rust
pub type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;
```

Ideally we should use our own error type.

**Comment by GopherJ** [comment]
The importance is to be `Send + Sync`

**Comment by GopherJ** [comment]
Let's track: https://github.com/jonathandturner/rhai/issues/129

**Comment by GopherJ** [comment]
closed by #91
